storage/innobase/dict/dict0stats.cc

```diff
@@ -114,12 +114,25 @@ where n=1..n_uniq.
 #define DEBUG_PRINTF(fmt, ...)	/* noop */
 #endif /* UNIV_STATS_DEBUG */
 
-/* Gets the number of leaf pages to sample in persistent stats estimation */
-#define N_SAMPLE_PAGES(index)					\
-	static_cast<ib_uint64_t>(				\
-		(index)->table->stats_sample_pages != 0		\
-		? (index)->table->stats_sample_pages		\
-		: srv_stats_persistent_sample_pages)
+/* Gets the number of leaf pages to sample in persistent stats estimation 
+   check if table's stats_percentage is 0
+     no > N_SAMPLE_PAGES calculation uses table's sample_percentage
+     yes  > check if table's sample_pages is 0
+       no > N_SAMPLE_PAGES is table's sample_pages number
+       yes > use instance level's sample_pages
+
+   The + 1 is to prevent int truncation after a division, without it the
+   instance could crash
+
+*/
+#define N_SAMPLE_PAGES(index)                               \
+    static_cast<ib_uint64_t>(                               \
+	(index)->table->stats_percentage != 0               \
+	? (index)->stat_index_size *                        \
+	(index)->table->stats_percentage / 100 + 1          \
+	: (index)->table->stats_sample_pages != 0           \
+	? (index)->table->stats_sample_pages                \
+	: srv_stats_persistent_sample_pages)
```

Comment:
Can this be defined as an `inline` function with some `if` and `else` instead of multiple ternary operators?

Reply:
Good idea. It was a define macro, but with a little bit of added logic, it's getting unwieldy. I'll change in a new commit that'll be pushed in today.